src/approaches/grammar_search_invention_approach.py

```diff
@@ -396,6 +397,220 @@ def _create_grammar(grammar_name: str, dataset: Dataset,
+        option_specs = [p.option_spec for p in partitions]
+        op_heuristic = self._evaluate(predicates, pruned_atom_data, segments,
+                                      strips_ops, option_specs)
+        pred_penalty = self._get_predicate_penalty(predicates)
+        op_penalty = self._get_operator_penalty(strips_ops)
+        total_score = op_heuristic + pred_penalty + op_penalty
+        print("Total score:", total_score)
+        return total_score
+
+    def _evaluate(self, predicates: FrozenSet[Predicate],
+                  pruned_atom_data: List[GroundAtomTrajectory],
+                  segments: List[Segment],
+                  strips_ops: List[STRIPSOperator],
+                  option_specs: List[OptionSpec]) -> float:
+        """Use learned operators to compute a heuristic for the predicates.
+        """
+        raise NotImplementedError("Override me!")
+
+    @staticmethod
+    def _get_operator_penalty(strips_ops: Collection[STRIPSOperator]
+                             ) -> float:
+        """Get a heuristic penalty based on the operator complexities.
+        """
+        size = 0
+        for op in strips_ops:
+            size += len(op.parameters) + len(op.preconditions) + \
+                    len(op.add_effects) + len(op.delete_effects)
+        return CFG.grammar_search_size_weight * size
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class _PredictionErrorHeuristic(_OperatorLearningBasedHeuristic):
+    """Score a predicate set by learning operators and counting false positives.
+    """
+
+    def _evaluate(self, predicates: FrozenSet[Predicate],
+                  pruned_atom_data: List[GroundAtomTrajectory],
+                  segments: List[Segment],
+                  strips_ops: List[STRIPSOperator],
+                  option_specs: List[OptionSpec]) -> float:
+        del predicates, pruned_atom_data  # unused
+        num_true_positives, num_false_positives, _, _ = \
+            _count_positives_for_ops(strips_ops, option_specs, segments)
+        return CFG.grammar_search_false_pos_weight * num_false_positives + \
+               CFG.grammar_search_true_pos_weight * (-num_true_positives)
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class _HAddBasedHeuristic(_OperatorLearningBasedHeuristic):
+    """Score a predicate set by learning operators and computing the hAdd
+    heuristic for the demonstration data.
+    """
+
+    def _evaluate(self, predicates: FrozenSet[Predicate],
+                  pruned_atom_data: List[GroundAtomTrajectory],
+                  segments: List[Segment],
+                  strips_ops: List[STRIPSOperator],
+                  option_specs: List[OptionSpec]) -> float:
+        score = 0.0  # lower is better
+        for traj, atoms_sequence in pruned_atom_data:
+            if not traj.is_demo:  # we only care about demonstrations
+                continue
+            # Create hAdd heuristic for this trajectory.
```

Comment:
this comment doesn't describe the entire rest of the for loop body, so add another comment below somewhere